manual/sphinx/developer_docs/file_io.rst

```diff
@@ -144,22 +151,116 @@ The important bits of the DataFormat interface are::
+
+      /// Gets an integer attribute
+      ///
+      /// Inputs
+      /// ------
+      ///
+      /// @param[in] varname     Variable name. The variable must already exist. If
+      ///                        varname is the empty string "" then get the
+      ///                        attribute from the top-level of the file instead
+      ///                        of from a variable.
+      /// @param[in] attrname    Attribute name
+      ///
+      /// Returns
+      /// -------
+      /// value                  An int attribute of the variable
+      virtual bool getAttribute(const std::string &varname, const std::string &attrname, int &value) = 0;
+
+      /// Gets a BoutReal attribute
+      ///
+      /// Inputs
+      /// ------
+      ///
+      /// @param[in] varname     Variable name. The variable must already exist. If
+      ///                        varname is the empty string "" then get the
+      ///                        attribute from the top-level of the file instead
+      ///                        of from a variable.
+      /// @param[in] attrname    Attribute name
+      ///
+      /// Returns
+      /// -------
+      /// value                  A BoutReal attribute of the variable
+      virtual bool getAttribute(const std::string &varname, const std::string &attrname, BoutReal &value) = 0;
+
+      /// Write out the meta-data of a field as attributes of the variable
+      void writeFieldAttributes(const std::string& name, const Field& f);
+      /// Overload for FieldPerp so we can also write 'yindex'
+      void writeFieldAttributes(const std::string& name, const FieldPerp& f);
+
+      /// Read the attributes of a field
+      void readFieldAttributes(const std::string& name, Field& f);
+      /// Overload for FieldPerp so we can also read 'yindex'
+      void readFieldAttributes(const std::string& name, FieldPerp& f);
     };
 
 .. [1] Support for PDB files was removed in BOUT++ 4.0.0
+
+FieldPerp I/O
+-------------
+
+`FieldPerp` objects can be saved to output files and read from them. The `yindex` of a
+`FieldPerp` is the local y-index on a certain processor, but is saved in output files as a
+global y-index in the attribute `yindex_global`. The intention is that a `FieldPerp` being
+saved should be a globally well-defined object, e.g. a set of values at one divertor
+target boundary, that will only be saved from processors holding that global y-index.
+Actually, the C++ I/O code should work fine even if a `FieldPerp` object is defined with
+different y-indices on different processors, but Python routines like `collect` and
+`restart.redistribute` will fail because they find inconsistent `yindex_global` values.
+
+`FieldPerp` objects should be added to the output files on every processor even if they
+are not allocated or used, otherwise `collect` cannot find the variable in the first
+output file to get its dimensions.
```

Comment:
Sorry, I don't quite understand this -- it seems inconsistent with the previous paragraph? If all processors write a `FieldPerp`, doesn't that mean that `yindex_global` will be different in each dump file, which will break `collect`? Or am I missing something?

Reply:
The idea is that there'll be a set of processors (connected in the x-direction) which all have a valid `FieldPerp` that they will write, all with the same `yindex_global`. All the other processors will write a `FieldPerp`, but with an invalid (i.e. sufficiently negative) `yindex_global`. Even invalid `FieldPerp`s in the dump files will have the right `bout_type` and dimensions, which is what `collect` needs from the variable in the first file.

I agree that second paragraph is not clear, I'll try and re-write. The idea is that you could write a FieldPerp from all processors (e.g. might be useful for some diagnostic/debugging purposes), but `collect` won't know what to do then (do you want `NXPE` `FieldPerp`s? if not which y-index do you want?). On the other hand, `collect` does need the `FieldPerp` variable to exist (even if invalid) in the first dump file, otherwise we'd have to search all dump files before we could say that a variable's missing. I guess that's possible, but would require more changes (and potentially be slow?).

Reply:
Ah, ok, so there should be two sets of `FieldPerps`: those on processes with identical `yindex_global`, and those with invalid values for `yindex_global`?